MREUnityRuntime/MREUnityRuntimeLib/Behaviors/Actions/Action.cs

```diff
@@ -80,27 +126,53 @@ public void StartAction(IUser user)
 		/// <param name="user">The user stopping the action.</param>
 		public void StopAction(IUser user)
 		{
-			UpdateAction(user, ActionState.Stopped);
+			ChangeState(user, ActionState.Stopped, null);
 		}
 
-		private void UpdateAction(IUser user, ActionState newState)
+		/// <summary>
+		/// Provides an action update while the action is being performed.
+		/// </summary>
+		/// <param name="user">The user performing the action.</param>
+		public void PerformActionUpdate()
 		{
-			if (user == null)
-			{
-				throw new ArgumentNullException("User cannot be null when performing an action");
-			}
+			UpdateAction(null);
+		}
+	}
 
-			ActionState oldState = ActionState.Stopped;
-			if (_userActionStates.TryGetValue(user, out oldState))
-			{
-				_userActionStates[user] = newState;
-			}
-			else
-			{
-				_userActionStates.Add(user, newState);
-			}
+	/// <summary>
+	/// The class that serves as the basic actions with action data that are a part of a behavior.
+	/// </summary>
+	/// <typeparam name="ActionDataT">The action data type associated with the action events.</typeparam>
+	public sealed class MWAction<ActionDataT> : MWActionBase
+		where ActionDataT : BaseActionData
+	{
+		/// <summary>
+		/// Signals the start of this action for the given user.
+		/// </summary>
+		/// <param name="user">The user starting the action.</param>
+		/// <param name="actionData">The optional data to pass along.</param>
+		public void StartAction(IUser user, ActionDataT actionData = null)
+		{
+			ChangeState(user, ActionState.Started, actionData);
+		}
 
-			Handler?.HandleActionStateChanged(user, oldState, newState);
+		/// <summary>
+		/// Signals the stop of this action for the given user.
+		/// </summary>
+		/// <param name="user">The user stopping the action.</param>
+		/// <param name="actionData">The optional data to pass along.</param>
+		public void StopAction(IUser user, ActionDataT actionData = null)
+		{
+			ChangeState(user, ActionState.Stopped, actionData);
```

Comment:
>ChangeState [](start = 3, length = 11)

I hope this question makes sense. When StopAction is called for the pen behavior, does it send the pen data accumulated since the last SynchronizeComponent call? Does it then clear out that data, or alternatively clear it on the next StartAction?

Reply:
Good catch.  Added in an assurance that the last perform action will be called before the stopping action is sent by the tool.